ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,563 @@
+   */
+  private class ClusterChangeListener extends RoutingTableProvider implements LiveInstanceChangeListener {
+    final Set<String> allInstances = new HashSet<>();
+    private boolean liveInstanceChangeTriggered = false;
+    private boolean externalViewChangeTriggered = false;
+    private boolean configChangeTriggered = false;
+    private CountDownLatch initialized = new CountDownLatch(3);
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      logger.info("Live instance change triggered with: " + liveInstances);
+      Set<String> liveInstancesSet = new HashSet<>();
+      for (LiveInstance liveInstance : liveInstances) {
+        liveInstancesSet.add(liveInstance.getInstanceName());
+        instanceNameToDataNodeIdMap.get(liveInstance.getInstanceName()).setState(HardwareState.AVAILABLE);
+      }
+      Set<String> downInstancesSet = Sets.difference(allInstances, liveInstancesSet);
+      for (String instanceName : downInstancesSet) {
+        instanceNameToDataNodeIdMap.get(instanceName).setState(HardwareState.UNAVAILABLE);
+      }
+      if (!liveInstanceChangeTriggered) {
+        liveInstanceChangeTriggered = true;
+        initialized.countDown();
+      }
+      helixClusterManagerMetrics.liveInstanceChangeTriggerCount.inc();
+    }
+
+    @Override
+    public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
+      logger.info("ExternalView change triggered with: " + externalViewList);
+      super.onExternalViewChange(externalViewList, changeContext);
+
+      // No action taken at this time.
+
+      if (!externalViewChangeTriggered) {
+        externalViewChangeTriggered = true;
+        initialized.countDown();
+      }
+      helixClusterManagerMetrics.externalViewChangeTriggerCount.inc();
+    }
+
+    @Override
+    public void onConfigChange(List<InstanceConfig> configs, NotificationContext changeContext) {
+      logger.info("Config change triggered with: ", configs);
+      super.onConfigChange(configs, changeContext);
+
+      // No action taken at this time. Going forward, changes like marking partitions back to read-write will go here.
+
+      if (!configChangeTriggered) {
+        configChangeTriggered = true;
+        initialized.countDown();
+      }
+      helixClusterManagerMetrics.instanceConfigChangeTriggerCount.inc();
+    }
+
+    /**
+     * Wait until the first set of notifications come in for live instance change, external view change and config
+     * changes.
+     * @throws InterruptedException if the wait gets interrupted unexpectedly.
+     */
+    void waitForInitialization() throws InterruptedException {
+      initialized.await();
+    }
+  }
+
+  /**
+   * Class that stores all ZK related information associated with a datacenter.
+   */
+  private static class DcZkInfo {
```

Comment:
It is more of a struct and does not need access to the enclosing parent class. It is a good idea to make it static in such cases.